core/src/main/java/org/jruby/MetaClass.java

```diff
@@ -33,31 +33,75 @@
 import org.jruby.runtime.builtin.IRubyObject;
 
 public final class MetaClass extends RubyClass {
-    /** rb_class_boot (for MetaClasses) (in makeMetaClass(RubyClass))
-     * 
-     */
+
+    @Deprecated
     public MetaClass(Ruby runtime, RubyClass superClass, IRubyObject attached) {
+        this(runtime, superClass, (RubyBasicObject) attached);
+    }
+
+    /**
+     * rb_class_boot for meta classes ({@link #makeMetaClass(RubyClass)})
+     */
+    MetaClass(Ruby runtime, RubyClass superClass, RubyBasicObject attached) {
         super(runtime, superClass, false);
         this.attached = attached;
-        setClassIndex(superClass.getClassIndex()); // use same ClassIndex as metaclass, since we're technically still of that type
+        // use same ClassIndex as metaclass, since we're technically still of that type
+        setClassIndex(superClass.getClassIndex());
+        superClass.addSubclass(this);
+    }
+
+    @Override
+    public final IRubyObject allocate(){
+        throw runtime.newTypeError("can't create instance of virtual class");
     }
 
     @Override
     public boolean isSingleton() {
         return true;
     }
 
-    public final IRubyObject allocate(){
-        throw runtime.newTypeError("can't create instance of virtual class");
+    /**
+     * rb_make_metaclass
+     * @param superClass
+     * @return singleton-class for this (singleton) class
+     */
+    @Override
+    public RubyClass makeMetaClass(RubyClass superClass) {
+        MetaClass klass = new MetaClass(runtime, getSuperSingletonMetaClass(), this);
+        setMetaClass(klass);
+
+        // Foo.singleton_class.singleton_class: #<Class:#<Class:Foo>>
+        // #<Class:#<Class:Foo>>'s singleton_class == #<Class:#<Class:Foo>>
+        klass.setMetaClass(klass);
+
+        return klass;
     }
 
-    public IRubyObject getAttached() {
+    private RubyClass getSuperSingletonMetaClass() {
+        if (attached instanceof RubyClass) {
+            RubyClass superClass = ((RubyClass) attached).getSuperClass();
+            if (superClass != null) superClass = superClass.getRealClass();
+            // #<Class:BasicObject>'s singleton class == Class.singleton_class
+            if (superClass == null) return runtime.getClassClass().getSingletonClass();
+            return superClass.getMetaClass().getSingletonClass();
+        }
+
+        return getSuperClass().getRealClass().getMetaClass(); // NOTE: is this correct?
+    }
+
+    @Override
+    RubyClass toSingletonClass(RubyBasicObject target) {
+        return attached == target ? this : super.toSingletonClass(target);
+    }
+
+    public RubyBasicObject getAttached() {
```

Comment:
ah, so I initially put in a `RubyClass` type there since the **attached** wasn't clear to be (almost) always a class.
not a must, but this is pretty much internal API, right? ext should pretty much see `RubyClass` as all there is ...

Reply:
its pretty much assumed to be of a RubyBasicObject

Reply:
@kares I was wondering about Module#initialize_copy.  It seems that can be RubyModule and not RubyClass.  Does that work here? like:

 ```ruby
module Foo; end; p Foo.send(:initialize_copy, Foo).singleton_class
```

As a broader question tracing through all this logic I half wonder whether rubymodule or rubyclass if we can narrow this more than basicobject (I see undef is basicobject but we could make that module/class maybe?).

Reply:
yeah, that "sinleton-class-clone-and-attache" code I wasn't sure ... 
didn't touch it - was hoping to get back to it, seems to be working : 
```
kares@clevo:~/workspace/oss/jruby$ bin/jruby -e "module Foo; end; p Foo.send(:initialize_copy, Foo).singleton_class"
#<Class:Foo>
```
```
kares@clevo:~/workspace/oss/jruby$ bin/jruby -e "module Foo; end; p Foo.send(:initialize_copy, Foo).singleton_class.singleton_class"
#<Class:#<Class:Foo>>
```
... super-class of such also seems fine (working as MRI) : 
```
kares@clevo:~/workspace/oss/jruby$ bin/jruby -e "module Foo; end; p Foo.send(:initialize_copy, Foo).singleton_class.singleton_class.superclass"
#<Class:Module>
```

at this point I am guessing only class singleton_class-es were really broken - which is what has been handled, thus hopefully this is the right direction.

Reply:
> half wonder whether rubymodule or rubyclass if we can narrow this more than basicobject (I see undef is basicobject but we could make that module/class maybe?)

yeah - except `UNDEF` that it should always be a `RubyModule` (or `RubyClass/MetaClass`)
did not try not using `UNDEF` ... it might work with passing `null` around :see_no_evil: 

and if its to be a "public" API - maybe smt to go along in `RubyClass`' `isSingleton()` to retrieve the "attached" class (not sure about naming) ... but that is mostly just throwing in random ideas :)

Reply:
@kares yeah UNDEF I don't know how that is used in this context either.  Maybe null works.  At some point we probably want to more formally restrict what attached means and what it can be.  RubyBasicObject pretty only eliminates a few things I don't think it could ever be already.

Reply:
actually, it might be fine as is ... was concentrating too much on class's singleton_class
in case of a object ... attaches the bare (BasicObject) object for `object.singleton_class`

that `null` case (instead of UNDEF) still seems legit, as there's some (old) TODO notes down the `getAttached()` path of dealing with NPEs ... which UNDEF will throw as it gets used.

Reply:
@kares I thought it might just attach for object singletons but I did not see which code path did that so I was confused.  I think perhaps even thinking that the name 'attached' probably does not best reflect what the field is for.  So another thing to ponder...a better name.

Reply:
open to suggestions here - already biased as got used to the attached name :)
anyway, should be good to go as is, we could iterate some more over later ...
okay for 9.2.1 or should we up hold till release?

Reply:
Yeah my comment was not directed to this PR really.  I just realized attached is probably the most minimal description possible.  'singletonFor'?  Anyways I think we can land this.  @headius does this worry you at all?